src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -149,25 +149,34 @@ private Task<Releaser> LockWaitingHelper(Task waitTask)
         {
             Requires.NotNull(waitTask, nameof(waitTask));
 
-            return waitTask.IsCompleted
-                ? this.uncontestedReleaser // uncontested lock
-                : waitTask.ContinueWith(
-                    (waiter, state) =>
-                    {
-                        // Re-throw any cancellation or fault exceptions.
-                        waiter.GetAwaiter().GetResult();
-
-                        if (this.disposed)
+            switch (waitTask.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    // uncontested lock
+                    return this.uncontestedReleaser;
+
+                case TaskStatus.Canceled:
+                    return this.canceledReleaser;
```

Comment:
@AArnott Should we update this line to preserve the cancellation token in the cancelled task, similar to what we do in the caller?

Reply:
I wouldn't normally care too much about that, but HeeJae has repeatedly called out places where he checks `OperationCanceledException.CancellationToken` so I guess people care about that sort of thing. But not in this particular fix, since the other method already does just this as well.